File: 09-06-2025_BST_With_Dead_End.java
import java.util.*;

class Node {
    int data;
    Node left, right;
    
    Node(int x) {
        data = x;
        left = right = null;
    }
}

class Solution {
    static class Triple {
        Node node;
        int min, max;

        Triple(Node node, int min, int max) {
            this.node = node;
            this.min = min;
            this.max = max;
        }
    }

    public boolean isDeadEnd(Node root) {
        if (root == null)
            return false;

        Queue<Triple> q = new LinkedList<>();
        q.add(new Triple(root, 1, (int)1e5));

        while (!q.isEmpty()) {
            Triple t = q.poll();
            Node cur = t.node;
            int min = t.min;
            int max = t.max;

            if (cur.left == null && cur.right == null && min == max) {
                return true;
            }

            if (cur.left != null) {
                q.add(new Triple(cur.left, min, cur.data - 1));
            }

            if (cur.right != null) {
                q.add(new Triple(cur.right, cur.data + 1, max));
            }
        }

        return false;
    }
}

File: 09-06-2025_BST_With_Dead_End.cpp
/*The Node structure is
class Node
{
    int data;
    struct Node *left;
    struct Node *right;

    Node(int x){
        data = x;
        left = NULL;
        right = NULL;
    }
};*/

class Solution {
  public:
    bool isDeadEnd(Node *root) {
        // Code here
        queue<pair<Node *, pair<int, int>>> q;
        q.push({root, {1, int(pow(10, 5))}});
        while(!q.empty())
        {
            Node * cur = q.front().first;
            int min = q.front().second.first;
            int max = q.front().second.second;
            q.pop();
            if(cur->left==NULL && cur->right==NULL)
            {
                if(min==max)
                {
                    return true;
                }
            }
            if(cur->left)
            {
                q.push({cur->left, {min, cur->data-1}});
            }
            if(cur->right)
            {
                q.push({cur->right, {cur->data+1, max}});
            }
        }
        return false;;
    }
};